CassetteTests/CASQueueFileTests.m

```diff
@@ -174,28 +177,48 @@ - (void)testAddingAndRemovingElementsWhileFragmentedAndFull {
     }
 }
 
-- (void)triggerStressedFragmentation {
+- (void)testRingReadingElementsMaintainsCorrectness {
+    NSUInteger numElementsToStartWith = 10;
+    NSUInteger numElementsToRemove = numElementsToStartWith / 2;
+    NSArray<NSData *> * dataArray = [self triggerStressedFragmentation:numElementsToStartWith
+                                                   numElementsToRemove:numElementsToRemove];
+    NSArray<NSData *> *elements = [self.queueFile peek:self.queueFile.size];
+    for (NSUInteger i = 0; i < elements.count; i++) {
+        XCTAssertEqualObjects(elements[(i + numElementsToRemove) % numElementsToStartWith],
+                              dataArray[i]);
+    }
+}
+
+- (NSMutableArray<NSData *> *)triggerStressedFragmentation:(NSUInteger)numElementsToStartWith
+                                       numElementsToRemove:(NSUInteger)numElementsToRemove {
     // Add just enough elements that it almost fills the storage
     // 4096 - 16 = 4080
     // Assumption: initial file size is 4096 - headerlength(16) = 4080
     // Assumption: element header length = 4
     NSUInteger initialFileLength = 4080;
-    NSUInteger numElementsToStartWith = 10;
     NSUInteger elementHeaderLength = 4;
-    NSUInteger sizeOfEachElement = initialFileLength / numElementsToStartWith;
-    NSData *data = [NSMutableData dataWithLength:(sizeOfEachElement - elementHeaderLength)];
+    NSUInteger sizeOfEachElement = (initialFileLength / numElementsToStartWith) - elementHeaderLength;
+    NSMutableArray<NSData *> *dataArray = [[NSMutableArray alloc] initWithCapacity:numElementsToStartWith];
+
+    // Add unique elements of data
     for (NSUInteger i = 0; i < numElementsToStartWith; i++) {
+        NSMutableData* data = [NSMutableData dataWithCapacity:sizeOfEachElement];
+        for (NSUInteger j = 0; j < sizeOfEachElement; j++) {
+            [data appendBytes:&i length:1];
+        }
         [self.queueFile add:data];
+        [dataArray addObject:data];
     }
 
     // At this point, buffer should be completely full.
```

Comment:
This fix seems fine to me, my only concern here is that we are waiting until the buffer is completely full to increase size of queue. My thought is it makes more sense to double size of queue at 75% rather than after we have reached max size. Let's talk offline about this and make it a separate task if we think changing it is is beneficial. thanks.